src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/AbstractReadThreadingGraph.java

```diff
@@ -465,6 +470,88 @@ private int recoverDanglingTail(final MultiDeBruijnVertex vertex, final int prun
         return mergeDanglingTail(danglingTailMergeResult);
     }
 
+    /**
+     * Finds the index of the best extent of the prefix match between the provided paths, for dangling head merging.
+     * Requires that at a minimum there are at least #getMinMatchingBases() matches between the reference and the read
+     * at the end in order to emit an alignmet offset.
+     *
+     * @param cigarElements cigar elements corresponding to the alignment between path1 and path2
+     * @param path1  the first path
+     * @param path2  the second path
+     * @return an integer pair object where the key is the offset into path1 and the value is offset into path2 (both -1 if no path is found)
+     */
+    private Pair<Integer, Integer> bestPrefixMatch(final List<CigarElement> cigarElements, final byte[] path1, final byte[] path2) {
+        final int getMinMatchingBases = getMinMatchingBases();
+        int matchesSinceLastMismatch = 0;
+        int refIndex = 0, readIdx = 0;
+        int bestRefIndex = 0, bestReadIdx = 0;
+        // we want to paste the base before the last read base, thus we need to handle it in the next iteration when we find a mismatch
+        boolean wasLastIndexAMismatch = false;
+
+        for (final CigarElement ce : cigarElements) {
```

Comment:
It looks like this can be made much simpler by traversing from the _end_ of the paths.  I think the pseudocode in this case becomes:

* go backwards from the right until you hit a mismatch or indel.
* if this covered fewer than `getMinMatchingBases()` return (-1,-1).
* otherwise return the read and ref position you got to.

In code this looks something like
```
refIdx = path1.length - 1;
readIdx = path2.length - 1;
for (final CigarElement ce : Lists.reverse(cigarElements)) {
   if (!(ce.consumesReadBases() && ce.consumesRefBases()) {
      break;
   }
   for (int j = 0; j < elementLength; j++, refIdx—; readIndx—) {
      if (path1[refIdx] != path2[readIdx]) {
         break;
      }
   }  
}

final int matches = path2.length - readIdx;
return matches < minimum ? (-1,-1) : (readIdx, refIdx);
```

Reply:
Right, your previous comment got lost in the rebase. I just checked and there is a problem with this, namely that there is no guarantee that the ref path and read path match in length. Thus in order to match the bases at the end of the dangling path I have at the very least parse through the cigar once forwards and then start walking it backwards... Perhaps that is still cleaner than what I wrote though...

Reply:
It looks like we excise the trailing deletions that might be useable for this at on this line: `        return new DanglingChainMergeHelper(altPath, refPath, altBases, refBases, AlignmentUtils.removeTrailingDeletions(alignment.getCigar()));`